# 1. Introduction ¶

As a data analyst, your job is to analyze data to extract valuable insights and make decisions accordingly. This involves going through different stages, such as data overview, preprocessing, and hypothesis testing.

Whenever we conduct research, we need to formulate hypotheses that we can then test. Sometimes we will accept these hypotheses; other times, we will reject them. To make the right decisions, a company must be able to understand whether it is making the right assumptions.

In this project, you will compare the music preferences of the cities of Springfield and Shelbyville. You will study real-world online music data to test the hypothesis below and compare user behavior in these two cities.

1.1. Objective:   
Test the hypothesis:   

User activity differs by day of the week and city.

1.2. Stages   
User behavior data is stored in the /datasets/music_project_en.csv file. There is no information about data quality, so you'll need to examine it before testing your hypothesis.

First, you should assess data quality and see if the issues are significant. Later, during data preprocessing, you'll need to address the most critical issues.

Your project will have these three stages:

- Data description.
- Data preprocessing.
- Hypothesis testing.

# Step 1. Data Description
Open the data and examine it.

Step 1.1. You'll need pandas, so import it.

In [1]:
# Import pandas
import pandas as pd

Step 1.2. Read the music_project_en.csv file from the /datasets/ folder and save it to the df variable:

In [2]:
# Read the file and store it in df
df = pd.read_csv("../data/raw/music_project_en.csv")

Step 1.3. Display the first 10 rows of the table:

In [3]:
# Get the first 10 rows from the df table
print(f"Sampling the first ten values ​​from the DataFrame: \n\n{df.head(10)}")

Sampling the first ten values ​​from the DataFrame: 

     userID           Track          artist   genre       City        time        Day
0  FFB692EC  Kamigata To...  The Mass Mi...    rock  Shelbyville  20:28:33  Wednesday
1  55204538  Delayed Bec...  Andreas Rön...    rock  Springfield  14:07:09     Friday
2    20EC38  Funiculì fu...     Mario Lanza     pop  Shelbyville  20:58:07  Wednesday
3  A3DD03C9  Dragons in ...      Fire + Ice    folk  Shelbyville  08:37:09     Monday
4  E2DC1FAE     Soul People      Space Echo   dance  Springfield  08:34:34     Monday
5  842029A1          Chains        Obladaet  rusrap  Shelbyville  13:09:41     Friday
6  4CB90AA5            True    Roman Messer   dance  Springfield  13:00:07  Wednesday
7  F03E1C1F  Feeling Thi...  Polina Grif...   dance  Springfield  20:47:49  Wednesday
8  8FA1D3BE        L’estate     Julia Dalia  ruspop  Springfield  09:17:40     Friday
9  E772D5C0       Pessimist             NaN   dance  Shelbyville  21:20:49  Wednesday


Step 1.4. Get general information about the table using the info() method.

In [4]:
# Get general information about our data
print("General Data Frame Information: \n")
df.info()

General Data Frame Information: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


These are our observations about the table. It contains seven columns that store the same data type: object.

According to the documentation:

- 'userID': user identifier;
- 'Track': song title;
- 'artist': artist name;
- 'genre': song genre;
- 'City': user city;
- 'time': the exact time the song was played;
- 'Day': day of the week.

We can see three issues with the style in the table headers:

1. Some headers are uppercase, others are lowercase.
2. There are spaces in some headers.
3. There is no consistent format; a standard naming format like snake_case should be used.

1.5. Write down some of your observations. Answer the following questions:
1. What type of data is in the rows? How can we know what the columns store?

2. Is there enough data to provide answers to our hypothesis, or do we need more information?

3. Did you notice any problems in the data, such as missing values, duplicates, or incorrect data types?

Write your answers here:

According to the data frame's .info() method, the data in the rows is of type object for the data frame (column = Dtype). You can see that the information provided by the .info() method and the .head() method is stored in the columns. You can see that string values ​​and NaN values ​​are stored in the columns.

There are 65,079 rows with 7 columns; not all cells contain relevant values, according to the count of non-Null values ​​in the columns. The hypothesis refers to user activity. I assume that the values ​​in the (time) and (day) columns could provide us with information related to the expected activity in the hypothesis.

There are 65079 rows with 7 columns, not all cells contain relevant values ​​according to the count of non-Null values. The .info() method shows us that the columns contain None or NaN values, the little information we obtained from .head(10), at first glance, does not show duplicate values, so we will have to check if there are explicit and implicit duplicate values.

# Step 2. Data Preprocessing   
Your goal here is to prepare the data for analysis. The first step is to resolve any issues with the headers. Then we can move on to missing and duplicate values. Let's get started!

Let's correct the formatting in the table headers.

Header Style   
Step 2.1. Display the table headers (column names):

In [5]:
# Display the column names
print("DataFrame Headers: \n\n", df.columns)

DataFrame Headers: 

 Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Let's change the table headers following conventional stylistic rules:

All characters must be lowercase.   
Remove spaces.   
If the name has multiple words, use snake_case.

Previously, you learned an automatic way to rename columns. Let's apply it now.

Step 2.2. Use the for loop to iterate over the column names and make all characters lowercase. When you're done, display the table headers again:

In [7]:
# Loop through headers that makes everything lowercase
for clmn_nm in df.columns:
    df.rename(columns={clmn_nm: clmn_nm.lower()}, inplace=True)

print("Clearing Data Frame headers (lowercase):\n\n", df.columns)

Clearing Data Frame headers (lowercase):

 Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Step 2.3. Now, using the same method, remove the spaces at the beginning and end of the column names and display the column names again:

In [8]:
# Loop through headers that removes spaces
for clmn_nm in df.columns:
    df.rename(columns={clmn_nm: clmn_nm.strip()}, inplace=True)

print("Data Frame header stripping (no spaces):\n\n", df.columns)

Data Frame header stripping (no spaces):

 Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Step 2.4. We need to apply the snake_case rule to the userid column. It should be user_id. Rename this column and display the names of all columns when you're done.

In [9]:
# Rename the "userid" column
df.rename(columns={"userid": "user_id"}, inplace=True)
print("Cleaning Data Frame headers (format):\n\n", df.columns)

Cleaning Data Frame headers (format):

 Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Step 2.5. Check the result. Display the headers again:

In [10]:
# Comprueba el resultado: lista de encabezados
print("Encabezados del Data Frame:\n\n", df.columns)

Encabezados del Data Frame:

 Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Missing Values   
Step 2.6. First, find the number of missing values ​​in the table. You must use two methods to obtain the number of missing values.

In [11]:
# Calculate the number of missing values
print("Missing values ​​in DataFrame: \n\n", df.isna().sum())

Missing values ​​in DataFrame: 

 user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


Not all missing values ​​affect the investigation in the same way. For example, missing values ​​in 'track' and 'artist' aren't crucial to the analysis, as these data are more descriptive than analytical. Therefore, you can directly replace them with a default value such as the string 'unknown'.

However, missing values ​​for 'genre' can influence the comparison between Springfield and Shelbyville's music preferences. In a real-world scenario, it would be ideal to investigate why this data is missing and attempt to recover it. But in this project, we don't have that option, so you'll need to fill in those values ​​with a default value.

As we saw earlier in the lessons, the best way to do this is to create a list with the names of the columns that need replacing. Then, use this list to iterate over each column and perform the corresponding replacement.

Step 2.7. Replace the missing values ​​in the 'track', 'artist', and 'genre' columns with the string 'unknown'.

1. Create a list called columns_to_replace containing the names of the 'track', 'artist', and 'genre' columns.

2. Use a for loop to iterate over each column in columns_to_replace.

3. Inside the loop, replace the missing values ​​in each column with the string 'unknown'.

In [12]:
# Loop through headers, replacing missing values ​​with 'unknown'
columns_to_replace = ["track", "artist", "genre"]

for column in columns_to_replace:
    df[column] = df[column].fillna('unknown')

Step 2.8. Now check the result to ensure there are no missing values ​​to replace in the data set. To do this, count the missing values ​​once more.

In [13]:
# Count missing values
print("Cleaning missing values ​​in Data Frame ('unknown'):\n\n", df.isna().sum())

Cleaning missing values ​​in Data Frame ('unknown'):

 user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


Duplicates   
Step 2.9. Find the number of explicit duplicates in the table. Again, you must apply two methods to obtain the number of explicit duplicates.

In [14]:
# Count explicit duplicates
print(f"Number of explicit duplicates in DataFrame: {df.duplicated().sum()}")

Number of explicit duplicates in DataFrame: 3826


Step 2.10. Now, remove all duplicates. To do this, call the method that does exactly this.

In [15]:
# Remove explicit duplicates
df.drop_duplicates(inplace=True)

Step 2.11. Let's now check if we've successfully eliminated all the duplicates. Count the explicit duplicates once more to make sure you've eliminated them all:

In [16]:
# Check for duplicates again
print(f"Number of explicit duplicates in DataFrame: {df.duplicated().sum()}")

Number of explicit duplicates in DataFrame: 0


Now we want to get rid of implicit duplicates in the genre column. For example, a genre name can be spelled in several ways. Such errors can also affect the result.

Step 2.11. First, we need to display a list of unique genre names, in alphabetical order. To do this:

1. Extract the genre column from the DataFrame.
2. Call the method that will return all unique values ​​in the extracted column.

In [17]:
# Inspect unique genre names
print("Unique values ​​in DataFrame: \n\n", df["genre"].unique())

Unique values ​​in DataFrame: 

 ['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hip' 'jazz' 'postrock' 'latin'
 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental' 'rusrock'
 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'hiphop' 'drum' 'extrememetal' 'türkçe'
 'experimental' 'easy' 'metalcore' 'modern' 'argentinetango' 'o

Let's examine the list to identify implicit duplicates of the hip-hop genre, that is, misspelled names or variants that refer to the same musical genre.

Some of the duplicates you'll find are:

- hip
- hop
- hip-hop

To fix this, let's create a function called replace_wrong_genres() that will have two parameters:

- wrong_genres: a list of all the values ​​to be replaced.
- correct_genre: a string to use as the replacement value.

The goal of this function is to correct the values ​​in the 'genre' column of the df DataFrame, replacing each value in the wrong_genres list with correct_genre.

Step 2.13

1. Define a function called replace_wrong_genres() that takes two parameters: wrong_genres and correct_genre.
2. Inside the function, use a for loop to iterate over each value in the wrong_genres list.
3. In each iteration, access the 'genre' column of the DataFrame df and use the .replace() method to replace the incorrect value with the correct_genre.
4. Call the function and pass as arguments:
    - A list of implicit duplicates: ['hip', 'hop', 'hip-hop']
    - The replacement string: 'hiphop'

In [19]:
# Function to replace implicit duplicates
def replace_wrong_genres(wrong_genres, correct_genre):
    
    for gnr in wrong_genres:
        df["genre"] = df["genre"].replace(gnr, correct_genre)

Step 2.14. Now, call replace_wrong_genres() and pass it these arguments to remove the implicit duplicates (hip, hop, and hip-hop) and replace them with hiphop:

In [20]:
# Remove implicit duplicates
genre_implicit_duplicates = ["hip", "hop", "hip-hop"]
replace_wrong_genres(genre_implicit_duplicates, "hiphop")

Step 2.15. Make sure the duplicate names have been removed. Display the list of unique values ​​for the 'genre' column again:

In [21]:
# Check for implicit duplicates again
print("Unique values ​​in DataFrame: \n\n", df["genre"].unique())

Unique values ​​in DataFrame: 

 ['rock' 'pop' 'folk' 'dance' 'rusrap' 'ruspop' 'world' 'electronic'
 'unknown' 'alternative' 'children' 'rnb' 'hiphop' 'jazz' 'postrock'
 'latin' 'classical' 'metal' 'reggae' 'triphop' 'blues' 'instrumental'
 'rusrock' 'dnb' 'türk' 'post' 'country' 'psychedelic' 'conjazz' 'indie'
 'posthardcore' 'local' 'avantgarde' 'punk' 'videogame' 'techno' 'house'
 'christmas' 'melodic' 'caucasian' 'reggaeton' 'soundtrack' 'singer' 'ska'
 'salsa' 'ambient' 'film' 'western' 'rap' 'beats' "hard'n'heavy"
 'progmetal' 'minimal' 'tropical' 'contemporary' 'new' 'soul' 'holiday'
 'german' 'jpop' 'spiritual' 'urban' 'gospel' 'nujazz' 'folkmetal'
 'trance' 'miscellaneous' 'anime' 'hardcore' 'progressive' 'korean'
 'numetal' 'vocal' 'estrada' 'tango' 'loungeelectronic' 'classicmetal'
 'dubstep' 'club' 'deep' 'southern' 'black' 'folkrock' 'fitness' 'french'
 'disco' 'religious' 'drum' 'extrememetal' 'türkçe' 'experimental' 'easy'
 'metalcore' 'modern' 'argentinetango' 'old' 's

In [ ]:
# Check if there are still incorrect genders
print(df[df['genre'].isin(genre_implicit_duplicates)])

Empty DataFrame
Columns: [user_id, track, artist, genre, city, time, day]
Index: []


In [26]:
# Count how many times the corrected gender appears
print(df['genre'].value_counts().loc['hiphop'])

3056


Observations   
Write a brief summary of what you discovered while analyzing the data. Your response should identify the problems you detected, explain how you resolved them, and describe how these actions improved the quality of the analysis.

Stage 2.16. Describe it here.

- Column Header Issues:   
The column names did not follow a consistent format, containing spaces and capital letters, making them difficult to manipulate.

    Solution: Column names were standardized using lowercase letters, removing spaces, and applying snake_case formatting.

    Impact: This facilitated data access and manipulation, improving readability and reducing errors in the analysis.

- Missing Values:   
Null values ​​were found in the track, artist, and genre columns.

    Solution: Missing values ​​in these columns were replaced with the string 'unknown'.

    Impact: This maintained the integrity of the dataset and prevented errors during the analysis.

- Explicit Duplicates:   
Duplicate rows were detected in the dataset.

    Solution: Explicit duplicates were removed using the drop_duplicates() method.

    Impact: This prevented duplicates from affecting the analysis results.

- Implicit duplicates in the genre column:   
Variants of the hiphop genre (hip, hop, hip-hop) were found.

    Solution: A function was created to replace these variants with a single standard value, hiphop.

    Impact: This improved data consistency and allowed for more accurate analysis of music genres.

# Step 3 Hypothesis testing

Task: Compare User Behavior in the Two Cities

The hypothesis we want to test is that there are differences in the way users in Springfield and Shelbyville consume music. To analyze this, we will focus on data from three specific days of the week: Monday, Wednesday, and Friday.

The analysis will consist of comparing the number of songs played by users in each city on those days. This will allow us to observe possible patterns or differences in consumption habits between Springfield and Shelbyville.

To carry out this analysis, it is important to follow the split-apply-combine approach, which we discussed in the lesson. In this case:

- Split: We separate the data into groups based on city.
- Apply: Within each group, we count how many songs were played.
- Combine: We compile the results into a structure that allows us to easily compare both cities.

This procedure should be repeated independently for each of the three selected days. The final result should show the number of plays by city on each of those days.

A possible way to structure the code would be with the following expression:

_df.groupby(by='...')['...'].method()_

You'll need to fill in the corresponding arguments to group by city and count the songs played. This approach will give you a clear, comparative view of user behavior in both cities.

Step 3.1. Count how many songs were played in each city using the 'track' column as a reference.

In [28]:
# Count the songs played in each city
print(f"Song plays by city: \n\n{df.groupby(by='city')['track'].count()}")

Song plays by city: 

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


Briefly write down your observations about the results. What differences did you notice between Springfield and Shelbyville? What might explain these differences?

Step 3.2. Write your observations here.

In Shelbyville, 18,512 songs were played. In Springfield, 42,741 songs were played.

The difference could be due to the number of residents in each city or the number of people who like to listen to music.

Step 3.3.

Group the data by day of the week and count how many songs were played on Mondays, Wednesdays, and Fridays.

Use the same counting method as before, but now change the grouping column to focus on the day.

This will allow you to identify possible listening patterns based on the day of the week.

In [30]:
# Calcula las canciones reproducidas en cada uno de los tres días
print(f"Reproduccion de canciones por dia: \n\n{df.groupby(by='day')['track'].count()}")
print()
print(f"Reproduccion de canciones por dia y ciudad: \n\n{df.groupby(by=['day', 'city'])['track'].count()}")

Reproduccion de canciones por dia: 

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

Reproduccion de canciones por dia y ciudad: 

day        city       
Friday     Shelbyville     5895
           Springfield    15945
Monday     Shelbyville     5614
           Springfield    15740
Wednesday  Shelbyville     7003
           Springfield    11056
Name: track, dtype: int64


Briefly describe what you observed when comparing Mondays, Wednesdays, and Fridays. Were there any days with less activity? Do the conclusions change if you analyze each city separately?

Step 3.4. Write your observations here.

- Analyzing the number of songs played in both cities, "Mondays" and "Fridays" are the days with the most music listening, while Wednesdays have the least amount of music. "Fridays" is the day with the most music listening.
- Analyzing the number of songs played by city and day, for SHELBYVILLE, "Wednesdays" and "Fridays" are the days with the most music listening, while Mondays have the least amount of music listening. "Mondays" is the day with the most music listening.
- Analyzing the number of songs played by city and day, for SPRINGFIELD, "Mondays" and "Fridays" are the days with the most music listening, while "Wednesdays" are the days with the least music listening. "Friday" is the day with the most music listening.
- Springfield is the city with the most music listening compared to Shelbyville.

So far, you've learned how to count entries by grouping them by a single criterion, such as city or day of the week. Now let's take it a step further: you need to create a function that counts how many songs were played in a specific city on a given day, combining both filtering criteria.

The function will be called number_tracks() and must accept two parameters:

- day: a day of the week (e.g., 'Monday').
- city: the name of a city (e.g., 'Springfield').

Within the function, you'll need to apply sequential filtering using logical indexing: first, you'll need to filter the DataFrame by day, and then by city. Once you have the correct subset of data, you'll need to count how many times a value appears in the 'user_id' column. This number will represent the total number of songs played under those two criteria.

Step 3.5

1. Declare a function called number_tracks() with two parameters: day and city.
2. Filter the DataFrame to keep only the rows where the 'day' column is equal to the value of the day parameter.
3. From the previous result, filter again to keep only the rows where the 'city' column is equal to the value of the city parameter.
4. Extract the 'user_id' column from the filtered DataFrame and use the .count() method to count the number of entries.
5. Store this value in a variable and return it as the function result.

In [31]:
# Declare the number_tracks() function with two parameters: day= and city=.
def number_tracks(dy, cty):

    # Stores the rows of the DataFrame where the value in the 'day' column is equal to the day= parameter.
    df_day_filtered = df[df["day"] == dy]

    # Filters the rows where the value in the 'city' column is equal to the city= parameter.
    df_city_filtered = df_day_filtered[df_day_filtered["city"] == cty]

    # Extracts the 'user_id' column from the filtered table and applies the count() method.
    df_user_id_count = df_city_filtered["user_id"].count()

    # Returns the number of values ​​in the 'user_id' column.
    return df_user_id_count

Step 3.6. Call number_tracks() six times, changing the parameter values ​​so that you can retrieve data for both cities for each of the three days.

In [32]:
# The number of songs played in Springfield on Monday
tracks = number_tracks("Monday", "Springfield")
print(f"The number of songs played in Springfield on Monday: {tracks}")

The number of songs played in Springfield on Monday: 15740


In [33]:
# The number of songs played in Shelbyville on Monday
tracks = number_tracks("Monday", "Shelbyville")
print(f"The number of songs played in Shelbyville on Monday: {tracks}")

The number of songs played in Shelbyville on Monday: 5614


In [34]:
# The number of songs played in Springfield on Wednesday
tracks = number_tracks("Wednesday", "Springfield")
print(f"The number of songs played in Springfield on Wednesday: {tracks}")

The number of songs played in Springfield on Wednesday: 11056


In [35]:
# The number of songs played in Shelbyville on Wednesday
tracks = number_tracks("Wednesday", "Shelbyville")
print(f"The number of songs played in Shelbyville on Wednesday: {tracks}")

The number of songs played in Shelbyville on Wednesday: 7003


In [36]:
# The number of songs played in Springfield on Friday
tracks = number_tracks("Friday", "Springfield")
print(f"The number of songs played in Springfield on Friday: {tracks}")

The number of songs played in Springfield on Friday: 15945


In [37]:
# The number of songs played in Shelbyville on Friday
tracks = number_tracks("Friday", "Shelbyville")
print(f"The number of songs played in Shelbyville on Friday: {tracks}")

The number of songs played in Shelbyville on Friday: 5895


# Conclusions   
## Write your final conclusions about the hypothesis.

Does the data support the idea that user behavior regarding music listening varies by city and day of the week?
Indicate whether the hypothesis should be accepted or rejected, and justify your answer based on your results.

Step 4. Detail your conclusions here.

Based on the analysis:

Springfield shows greater musical activity compared to Shelbyville.

The day of the week influences user behavior. For example, Fridays tend to have the most activity, while Wednesdays have the least.

The data support the hypothesis that user behavior varies by city and day of the week.

**Note**:   
In real-world research projects, statistical hypothesis testing is more precise and quantitative. Also, keep in mind that you can't always draw conclusions about an entire city from data from a single source.

You'll learn more about hypothesis testing in the Statistical Data Analysis Sprint.